### ISSUE TITLE
7.0.1: Activesupport instrumentor:  nil can't be coerced into Float (TypeError)

### ISSUE THREAD

**Issue by notapatch** [rails/rails#44167]
I have a [Rails 7.0.1 test application](https://github.com/notapatch/issue_rails_7_0_multiple_image_loading_error) that when it tries to load in multiple images (more than 2) it can regularly fail to load one or more images: on my dev system it happens most of the time.

Each time it fails you get the same 500 error:   `nil can't be coerced into Float (TypeError)` from activesupport instrumenter. This happens on Rails 7.0 only and not on the almost identical 6.1 version. 

### Rails 7.0.1 - fails to load one of the images most of the time.
![rails-7 0 1](https://user-images.githubusercontent.com/1710795/149349302-0b99ec27-d5d9-410a-9303-aa990865465c.png)

### Rails 6.1 - never seen if fail
![rails-6 1 4](https://user-images.githubusercontent.com/1710795/149348705-6025f791-7631-4af8-a4c3-80d7f58258b6.png)

### Stack trace

[Code that throws the exception is here](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/notifications/instrumenter.rb#L133)

```
Puma caught this error: nil can't be coerced into Float (TypeError)
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/activesupport-7.0.1/lib/active_support/notifications/instrumenter.rb:133:in `-'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/activesupport-7.0.1/lib/active_support/notifications/instrumenter.rb:133:in `duration'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/activesupport-7.0.1/lib/active_support/core_ext/enumerable.rb:61:in `map'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/activesupport-7.0.1/lib/active_support/core_ext/enumerable.rb:61:in `sum'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/activesupport-7.0.1/lib/active_support/core_ext/enumerable.rb:298:in `sum'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/actionpack-7.0.1/lib/action_dispatch/middleware/server_timing.rb:26:in `block in call'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/actionpack-7.0.1/lib/action_dispatch/middleware/server_timing.rb:25:in `each'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/actionpack-7.0.1/lib/action_dispatch/middleware/server_timing.rb:25:in `map'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/actionpack-7.0.1/lib/action_dispatch/middleware/server_timing.rb:25:in `call'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/actionpack-7.0.1/lib/action_dispatch/middleware/executor.rb:14:in `call'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/actionpack-7.0.1/lib/action_dispatch/middleware/static.rb:23:in `call'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/rack-2.2.3/lib/rack/sendfile.rb:110:in `call'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/actionpack-7.0.1/lib/action_dispatch/middleware/host_authorization.rb:137:in `call'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/railties-7.0.1/lib/rails/engine.rb:530:in `call'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/puma-5.5.2/lib/puma/configuration.rb:249:in `call'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/puma-5.5.2/lib/puma/request.rb:77:in `block in handle_request'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/puma-5.5.2/lib/puma/thread_pool.rb:340:in `with_force_shutdown'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/puma-5.5.2/lib/puma/request.rb:76:in `handle_request'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/puma-5.5.2/lib/puma/server.rb:447:in `process_client'
/Users/me/.asdf/installs/ruby/3.0.3/lib/ruby/gems/3.0.0/gems/puma-5.5.2/lib/puma/thread_pool.rb:147:in `block in spawn_thread'
```
### Steps to reproduce
 
I have a couple of apps one which works based on [6.1](https://github.com/notapatch/issue_rails_6_1_multiple_image_loading_error) and another that doesn't work based on [7.0.1](https://github.com/notapatch/issue_rails_7_0_multiple_image_loading_error)

1. Clone repo
2. cd into rails directory
3. bin/setup  (standard setup except it runs db:seed)
4. Repeat the following steps until you notice an image not loading
    1. bin/rails server
    2. browser opens localhost:3000 and keep "Network" tab open
    3. check if the 10 images all loaded
    4. close rails server  and back to step i  (It's important to stop server as normally once images loads they keep loading)
5. Check network tab for the image that failed to load and look at the response

### Browser Network tab showing exceptions
![network-tab-showing-exception](https://user-images.githubusercontent.com/1710795/149358866-d33484f8-9c12-46a4-89d2-6ef3d5662395.png)

There's very little code changes in the example applications. Basically:

1.  new application
2. scaffold product
3. install active_storage
4. product has_one_attached image
5. seed products with images
6. Start running the server following the steps to reproduce ...

I've also tried:
- With Rails 7.0 displaying static images from public - loads it fine.
- All the browsers show the same issue
- Rails 6.1 version works with or without spring but with spring every now and then it could hang until you did a `spring stop`.
- Larger images 300 x 300 also failed in similar manner and not noticeably more often.

### Expected behavior
It should load the images every time.

### Actual behavior
Most of the time in the test application it fails

### System configuration
Rails 7.0.1

Ruby 3.0.3

**Comment by AlexKovynev** [comment]
https://github.com/rails/rails/blob/main/actionpack/lib/action_dispatch/middleware/server_timing.rb

@sebastiansogamoso it seems not all events have a start time

**Comment by AlexKovynev** [comment]
And maybe try to use something like this cause it some more tricky https://github.com/dpogue/rails_server_timings
@rafaelfranca look like a serious bug. Unfortunately don't have enough knowledge of rails internals to suggest PR:(

**Comment by mihaic195** [comment]
I cannot reproduce the issue. Seems to work fine on my end.

**Comment by akpatternica** [comment]
I reproduce it easy with test application. Ubuntu 20.04 if it something change:)

**Comment by mihaic195** [comment]
Indeed, I've managed to reproduce. Can you give it a try with this PR? https://github.com/rails/rails/pull/44187

**Comment by akpatternica** [comment]
> Indeed, I've managed to reproduce. Can you give it a try with this PR? #44187

yep all fine, when propose a PR don't forget to ask port to 7.0 please:)

**Comment by mihaic195** [comment]
> > Indeed, I've managed to reproduce. Can you give it a try with this PR? #44187
> 
> yep all fine, when propose a PR don't forget to ask port to 7.0 please:) And your PR broke some tests it seems:(
> 
> But it seems it not your fault:)

Retriggered the build and now everything is green. Need to wait for approval.

**Comment by rafaelfranca** [comment]
We need to understand why an event doesn't have a start or finish time. Rails doesn't create any of those. Looking in the code, that logic didn't change in 12 years, so if you create an event without this information it would always raise. Does anyone have an application reproducing the problem?

**Comment by akpatternica** [comment]
@rafaelfranca  you can see it on the top of the issues https://github.com/notapatch/issue_rails_7_0_multiple_image_loading_error

**Comment by AlexKovynev** [comment]
@rafaelfranca 
ActiveSupport::Notifications::Fanout::Subscribers::Timed#finish(name="request.action_dispatch", id="d58d2936bd504758631f", payload={:request=>#<ActionDispatch::Request GET...) at ~/.rvm/gems/ruby-3.1.0/gems/activesupport-7.0.1/lib/active_support/notifications/fanout.rb:236

**Comment by jaynetics** [comment]
@rafaelfranca

this seems to be a bit of a bigger problem.

a new rails app will fail random requests in dev env because of this.

## minimal repro

1. `$ rails new testapp && cd testapp && bin/setup`

2. add this code:

```ruby
# in routes.rb:
root "application#index"

# in application_controller.rb:
class ApplicationController < ActionController::Base
  def index
    render html: 'ok', layout: true
  end
end
```

3. run `$ rails s` and open the page. check the console. sometimes random assets (e.g. js files from stimulus) will fail to load with the above error. (might need to restart the server a few times for this to happen)

## analysis

the broken event has `@name == "request.action_dispatch" and @time == nil`.

this happens because at the end of the request, there are [too many `@other_subscribers` in `Fanout`]( https://github.com/rails/rails/blob/0ea374c81f6b3af1e8adac6d6337a4ae54e8d73b/activesupport/lib/active_support/notifications/fanout.rb#L113 ), leading to [`Timed#finish` being called more often]( https://github.com/rails/rails/blob/0ea374c81f6b3af1e8adac6d6337a4ae54e8d73b/activesupport/lib/active_support/notifications/fanout.rb#L76 ) than `timestack.size`.

as a result, [`timestack.pop` will return `nil`]( https://github.com/rails/rails/blob/0ea374c81f6b3af1e8adac6d6337a4ae54e8d73b/activesupport/lib/active_support/notifications/fanout.rb#L235 ) for some of the emitted events.

the number of subscribers varies **a lot**, between 0 and 5. this does not happen with `$ RAILS_MAX_THREADS=1 rails s`, so i guess there is some mixup between subscribing and unsubscribing from different threads, in spite of the code comment that points out the thread-safety of `Fanout`...

this is an old problem by the way and not caused by recent changes like the addition of `IsolatedExecutionState`. i can reproduce it in rails 6.1 if i backport the `ServerTiming` middleware.

**Comment by jaynetics** [comment]
this is a bug in `Fanout`.

my impression is that `Fanout` should be refactored so as to keep the stacks per event and not per listener - but maybe @tenderlove remembers why he built it that way 8 years ago? :D

## most minimal repro

```ruby
events = []

10.times do
  Thread.new do
    sleep(rand(0.1))
    subscriber = ActiveSupport::Notifications.subscribe('foo') do |*args|
      events << ActiveSupport::Notifications::Event.new(*args)
    end
    ActiveSupport::Notifications.instrumenter.start('foo', {})

    sleep(rand(0.1))
    ActiveSupport::Notifications.instrumenter.finish('foo', {})
    ActiveSupport::Notifications.unsubscribe(subscriber)
  end
end

sleep(0.3)

# We have more than 10 events because many are caught by multiple subscribers
events.count # => 32

# Some events have no start time because listeners were added after the event
# start, leading to more #finish than #start calls and a depleted timestack.
events.count(&:time) # => 27
```

**Comment by jhawthorn** [comment]
Thank you all for the reproductions, they really helped 😁.

I think this is the combination of several issues: some limitations with the existing code and a new change which caused the combination to fail.

The change leading to this issue appearing in 7.0 is the `ActionDispatch::ServerTiming` middleware. Specifically the [dynamic subscription being done before and after each request](https://github.com/rails/rails/blob/main/actionpack/lib/action_dispatch/middleware/server_timing.rb#L15-L17). Doing a subscription like this has never really been safe throughout the history of AS::Notifications, and the [docs advise against it](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/notifications.rb#L127-L134) (though the language should be stronger, it's not just a performance issue). I also don't think this subscription works as intended if it wasn't for limitations of notifications. The topics are subscribed to globally so if we are serving requests in parallel this receives instrumentation from all active Threads. Additionally nested notifications even from the same request are going to be reported incorrectly. We probably need to revisit this implementation.

Why aren't dynamic subscriptions more of an issue? Usually we're able to avoid the issues that can cause by [using finish_with_state](https://github.com/rails/rails/blob/main/activesupport/lib/active_support/notifications/instrumenter.rb#L30) which ensures we're iterating over the same set of listeners. This is the path taken by `Notifications.instrument` and similar friendly user-facing APIs. However there's one place (that I found) where we're not: [Rack::Logger](https://github.com/rails/rails/blob/main/railties/lib/rails/rack/logger.rb#L70).

Because `Rack::Logger` uses `finish` directly it gets a new list of subscribers. If there had been subscriptions added in between the start and finish, we'll have too many subscribers (as mentioned above) and will be popping more than we should off the stack, leading to the `nil`s. Fortunately, we can solve this relatively simply by using `finish_with_state`. I've opened #44454, I'd love help verifying that this fixes the issues 🙇‍♂️.

I think this issue is good indication that we should be deprecating the "stateless" `start`/`finish` and ensure everything is using `start`/`finish_with_state` (or ideally just the block form of `instrument`).

Finally, I do think we should improve notifications to better handle this. Popping values off of a stack for each individual subscriber is slow and error prone. I spiked on avoiding all stack operations for the stateful case, which I'll open a PR for soon, but even in the stateless case if we had one stack item per-event rather than per event-subscriber this issue bug have been worked around.

---

TLDR: **For anyone running into this the best immediate fix is to disable server_timing in `development.rb`**. #44454 should solve the issue, and I hope to backport it to 7-0-stable. We should find another way to implement `ServerTiming`. Notifications could also use improvement.

**Comment by joker-777** [comment]
Hi, why is this closed when https://github.com/rails/rails/pull/44349 isn't merged yet? We are still experiencing this error with activesupport `7.0.3.1`.

**Comment by skipkayhil** [comment]
> Hi, why is this closed when #44349 isn't merged yet? We are still experiencing this error with activesupport `7.0.3.1`.

This was fixed in #44454 and backported to 7-0-stable, but just hasn't been released yet. In the meantime you can disable the ServerTiming middleware (or point your `Gemfile` at `7-0-stable`) and the issue should go away.

**Comment by akpatternica** [comment]
@skipkayhil is it really backported? I can't find it in after 7.0.3 commits.

**Comment by skipkayhil** [comment]
> is ported before 7.0.3

Ah, you're right, thank you. @joker-777 if you have a way to reproduce the issue on `7.0.3.1` feel free to open a new issue.

**Comment by joker-777** [comment]
@skipkayhil Unfortunately it is very hard for us to reproduce it reliably. Looking at the stack trace though it seems very similar to this error here:

```

Puma caught this error: nil can't be coerced into Float (TypeError)
/usr/local/bundle/gems/activesupport-7.0.3.1/lib/active_support/notifications/instrumenter.rb:133:in `-'
/usr/local/bundle/gems/activesupport-7.0.3.1/lib/active_support/notifications/instrumenter.rb:133:in `duration'
/usr/local/bundle/gems/activesupport-7.0.3.1/lib/active_support/core_ext/enumerable.rb:78:in `map'
/usr/local/bundle/gems/activesupport-7.0.3.1/lib/active_support/core_ext/enumerable.rb:78:in `sum'
/usr/local/bundle/gems/activesupport-7.0.3.1/lib/active_support/core_ext/enumerable.rb:315:in `sum'
/usr/local/bundle/gems/actionpack-7.0.3.1/lib/action_dispatch/middleware/server_timing.rb:26:in `block in call'
/usr/local/bundle/gems/actionpack-7.0.3.1/lib/action_dispatch/middleware/server_timing.rb:25:in `each'
/usr/local/bundle/gems/actionpack-7.0.3.1/lib/action_dispatch/middleware/server_timing.rb:25:in `map'
/usr/local/bundle/gems/actionpack-7.0.3.1/lib/action_dispatch/middleware/server_timing.rb:25:in `call'
/usr/local/bundle/gems/actionpack-7.0.3.1/lib/action_dispatch/middleware/executor.rb:14:in `call'
/usr/local/bundle/gems/actionpack-7.0.3.1/lib/action_dispatch/middleware/static.rb:23:in `call'
/usr/local/bundle/gems/rack-2.2.4/lib/rack/sendfile.rb:110:in `call'
/usr/local/bundle/gems/actionpack-7.0.3.1/lib/action_dispatch/middleware/host_authorization.rb:137:in `call'
/usr/local/bundle/gems/ddtrace-1.3.0/lib/datadog/tracing/contrib/rack/middlewares.rb:83:in `call'
/usr/local/bundle/gems/rack-mini-profiler-3.0.0/lib/mini_profiler/profiler.rb:249:in `call'
/usr/local/bundle/gems/railties-7.0.3.1/lib/rails/engine.rb:530:in `call'
/usr/local/bundle/gems/puma-5.6.4/lib/puma/configuration.rb:252:in `call'
/usr/local/bundle/gems/puma-5.6.4/lib/puma/request.rb:77:in `block in handle_request'
/usr/local/bundle/gems/puma-5.6.4/lib/puma/thread_pool.rb:340:in `with_force_shutdown'
/usr/local/bundle/gems/puma-5.6.4/lib/puma/request.rb:76:in `handle_request'
/usr/local/bundle/gems/puma-5.6.4/lib/puma/server.rb:441:in `process_client'
/usr/local/bundle/gems/puma-5.6.4/lib/puma/thread_pool.rb:147:in `block in spawn_thread'
```

**Comment by skipkayhil** [comment]
Perfect, that stack helped me narrow it down. I believe its the same issue that the PR above solved, but its in the `ddtrace` gem https://github.com/DataDog/dd-trace-rb/blob/cb1cfc1975ca6c70b5e9d76dc3b6f8272cf858c9/lib/datadog/tracing/contrib/rack/middlewares.rb#L112

I would open an issue in their repo with https://github.com/rails/rails/commit/3ab161632564b0980b2b8034f595b41d478cbf9b this example for how to implement this safely

**Comment by gingerlime** [comment]
Hi @skipkayhil ! thanks for your help. I work with @joker-777. I tested by removing `ddtrace` and still see errors on the backend some times

`Rack app ("GET /assets/app_javascript_components_quiz_type_selector_js-app_javascript_components_terminology_modal_j-e38558-chunk-e0c804251e29c02701b7bdac8d312af123767d992670295bc2c4d93dca570fa7.digested.js" - (91.*.*.*)): #<TypeError: nil can't be coerced into Float>`

The backend error doesn't have any trace info unfortunately, and I wasn't able to reproduce getting the error on the browser like posted above by @joker-777. As soon as I see the full trace, I would share it.

**Comment by gingerlime** [comment]
Finally got the full trace from Puma :)

```
Puma caught this error: nil can't be coerced into Float (TypeError)
/usr/local/bundle/gems/activesupport-7.0.3.1/lib/active_support/notifications/instrumenter.rb:133:in `-'
/usr/local/bundle/gems/activesupport-7.0.3.1/lib/active_support/notifications/instrumenter.rb:133:in `duration'
/usr/local/bundle/gems/activesupport-7.0.3.1/lib/active_support/core_ext/enumerable.rb:78:in `map'
/usr/local/bundle/gems/activesupport-7.0.3.1/lib/active_support/core_ext/enumerable.rb:78:in `sum'
/usr/local/bundle/gems/activesupport-7.0.3.1/lib/active_support/core_ext/enumerable.rb:315:in `sum'
/usr/local/bundle/gems/actionpack-7.0.3.1/lib/action_dispatch/middleware/server_timing.rb:26:in `block in call'
/usr/local/bundle/gems/actionpack-7.0.3.1/lib/action_dispatch/middleware/server_timing.rb:25:in `each'
/usr/local/bundle/gems/actionpack-7.0.3.1/lib/action_dispatch/middleware/server_timing.rb:25:in `map'
/usr/local/bundle/gems/actionpack-7.0.3.1/lib/action_dispatch/middleware/server_timing.rb:25:in `call'
/usr/local/bundle/gems/actionpack-7.0.3.1/lib/action_dispatch/middleware/executor.rb:14:in `call'
/usr/local/bundle/gems/actionpack-7.0.3.1/lib/action_dispatch/middleware/static.rb:23:in `call'
/usr/local/bundle/gems/rack-2.2.4/lib/rack/sendfile.rb:110:in `call'
/usr/local/bundle/gems/actionpack-7.0.3.1/lib/action_dispatch/middleware/host_authorization.rb:137:in `call'
/usr/local/bundle/gems/rack-mini-profiler-3.0.0/lib/mini_profiler/profiler.rb:237:in `call'
/usr/local/bundle/gems/railties-7.0.3.1/lib/rails/engine.rb:530:in `call'
/usr/local/bundle/gems/puma-5.6.4/lib/puma/configuration.rb:252:in `call'
/usr/local/bundle/gems/puma-5.6.4/lib/puma/request.rb:77:in `block in handle_request'
/usr/local/bundle/gems/puma-5.6.4/lib/puma/thread_pool.rb:340:in `with_force_shutdown'
/usr/local/bundle/gems/puma-5.6.4/lib/puma/request.rb:76:in `handle_request'
/usr/local/bundle/gems/puma-5.6.4/lib/puma/server.rb:441:in `process_client'
/usr/local/bundle/gems/puma-5.6.4/lib/puma/thread_pool.rb:147:in `block in spawn_thread'
```

So it seems unrelated to `ddtrace`

**Comment by jhawthorn** [comment]
It's likely your issue will be fixed via https://github.com/rails/rails/pull/44454, which is backported to 7.1 but has not yet been included in a release (we expect to have one shortly).

From your backtrace it's also clear the server_timing middleware is enabled. If you need an immediate fix disabling `server_timing` should solve your issue.

**Comment by AlexKovynev** [comment]
@jhawthorn it look like #44454 already included in 7.0.3

**Comment by jhawthorn** [comment]
Ah! You're right, sorry.

In that case if you can find a reproduction please file a new issue, I'd love to address it. It's likely that something is misbehaving by using `ActiveSupport::Notifications.finish` (rather than the preferred `Notifications.instrument` block or the less-preferred-but-acceptable `finish_with_state`).

But again, I would strongly recommend disabling the server timing middleware.
